File: mobile/src/main/java/treehou/se/habit/ui/settings/subsettings/general/GeneralSettingsPresenter.java
package treehou.se.habit.ui.settings.subsettings.general;

import com.f2prateek.rx.preferences2.Preference;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import treehou.se.habit.module.RxPresenter;
import treehou.se.habit.util.Settings;

public class GeneralSettingsPresenter extends RxPresenter implements GeneralSettingsContract.Presenter {

    @Inject Settings settings;

    private GeneralSettingsContract.View view;

    @Inject
    public GeneralSettingsPresenter(GeneralSettingsContract.View view) {
        this.view = view;
    }


    @Override
    public void subscribe() {
        super.subscribe();

        Observable<Boolean> settingsAutoloadSitemapRx = settings.getAutoloadSitemapRx();
        settingsAutoloadSitemapRx
                .compose(bindToLifecycle())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(show -> view.showAutoLoadSitemap(show));

        Preference<Boolean> settingsShowSitemapInMenuRx = settings.getShowSitemapsInMenuRx();
        settingsShowSitemapInMenuRx.asObservable()
                .compose(bindToLifecycle())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(show -> view.showSitemapsInMenu(show));

        Preference<Boolean> settingsFullscreenRx = settings.getFullscreenPref();
        settingsFullscreenRx.asObservable()
                .compose(bindToLifecycle())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(show -> view.setFullscreen(show));
    }

    @Override
    public void themeSelected(int theme) {
        Preference<Integer> themePref = settings.getThemePref();
        if(themePref.get() != theme) {
            themePref.set(theme);
            view.updateTheme();
        }
    }

    @Override
    public void setShowSitemapsInMenu(Boolean show) {
        Preference<Boolean> showSitemapsInMenuRx = settings.getShowSitemapsInMenuRx();
        if(show != showSitemapsInMenuRx.get()) {
            showSitemapsInMenuRx.set(show);
        }
    }

    @Override
    public void setAutoLoadSitemap(boolean show) {
        Observable<Boolean> autoloadSitemapRx = settings.getAutoloadSitemapRx();
        if(show != autoloadSitemapRx.blockingFirst()) {
            settings.setAutoloadSitemapRx(show);
        }
    }

    @Override
    public void setFullscreen(boolean fullscreen) {
        Preference<Boolean> fullscreenPrefRx = settings.getFullscreenPref();
        if(fullscreen != fullscreenPrefRx.get()) {
            fullscreenPrefRx.set(fullscreen);
        }
    }
}

File: mobile/src/main/java/treehou/se/habit/ui/settings/subsettings/general/GeneralSettingsPresenter.kt
package treehou.se.habit.ui.settings.subsettings.general

import io.reactivex.android.schedulers.AndroidSchedulers
import treehou.se.habit.dagger.RxPresenter
import treehou.se.habit.util.Settings
import treehou.se.habit.util.logging.Logger
import javax.inject.Inject

class GeneralSettingsPresenter @Inject
constructor(private val view: GeneralSettingsContract.View) : RxPresenter(), GeneralSettingsContract.Presenter {

    @Inject lateinit var settings: Settings
    @Inject lateinit var logger: Logger


    override fun subscribe() {
        super.subscribe()

        val settingsAutoloadSitemapRx = settings.autoloadSitemapRx
        settingsAutoloadSitemapRx
                .compose(bindToLifecycle())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ show -> view.showAutoLoadSitemap(show!!) }
                        , { logger.e(TAG, "settingsAutoloadSitemapRx update failed", it) })

        val settingsShowSitemapInMenuRx = settings.showSitemapsInMenuRx
        settingsShowSitemapInMenuRx.asObservable()
                .compose(bindToLifecycle())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe ({ show -> view.showSitemapsInMenu(show) }
                        , { logger.e(TAG, "settingsShowSitemapInMenuRx update failed", it) })

        val settingsFullscreenRx = settings.fullscreenPref
        settingsFullscreenRx.asObservable()
                .compose(bindToLifecycle())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe ({ show -> view.setFullscreen(show!!) }
                        , { logger.e(TAG, "settingsFullscreenRx update failed", it) })
    }

    override fun themeSelected(theme: Int) {
        val themePref = settings.themePref
        if (themePref.get() != theme) {
            themePref.set(theme)
            view.updateTheme()
        }
    }

    override fun setShowSitemapsInMenu(show: Boolean) {
        settings.showSitemapsInMenuRx.set(show)
    }

    override fun setAutoLoadSitemap(show: Boolean) {
        val autoloadSitemapRx = settings.autoloadSitemapRx
        if (show != autoloadSitemapRx.blockingFirst()) {
            settings.setAutoloadSitemapRx(show)
        }
    }

    override fun setFullscreen(fullscreen: Boolean) {
        val fullscreenPrefRx = settings.fullscreenPref
        if (fullscreen != fullscreenPrefRx.get()) {
            fullscreenPrefRx.set(fullscreen)
        }
    }

    companion object {
        val TAG = "GeneralSettingsPresenter"
    }
}
